feat(useScrollParent): easily get the nearest scrollable parent

yue1123 opened the pull request:
easily get the nearest scrollable parent reactively.  
<!-- Thank you for contributing! -->

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vueuse/vueuse/blob/main/packages/guidelines.md).
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

<details>
<summary><strong>⚠️ Slowing down new functions</strong></summary>
<br>

> **Warning**: **Slowing down new functions**
>
> As the VueUse audience continues to grow, we have been inundated with an overwhelming number of feature requests and pull requests. As a result, maintaining the project has become increasingly challenging and has stretched our capacity to its limits. As such, in the near future, we may need to slow down our acceptance of new features and prioritize the stability and quality of existing functions. **Please note that new features for VueUse may not be accepted at this time.** If you have any new ideas, we suggest that you first incorporate them into your own codebase, iterate on them to suit your needs, and assess their generalizability. If you strongly believe that your ideas are beneficial to the community, you may submit a pull request along with your use cases, and we would be happy to review and discuss them. Thank you for your understanding.

</details>

---

### Description

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->

yue1123 committed:
feat(useScrollParent): easily get the nearest scrollable parent reactively

ferferga commented on packages/core/useScrollParent/index.ts:
```diff
@@ -0,0 +1,87 @@
+    return root
+  }
+
+  const update = () => {
```
Any reason why this is not computed?

yue1123 replied:
There doesn't really seem to be a reason not to use computed.

yue1123 committed:
feat(useScrollParent): refactor to computed

yue1123 committed:
feat(useScrollParent): use document.scrollingElement as fallback and improve types

ferferga commented on packages/core/useScrollParent/index.ts:
```diff
@@ -0,0 +1,81 @@
+    lookupLevel = 10,
+    onError = (e) => { console.error(e) },
+  } = options
+  const overflowScrollReg = /scroll|auto|overlay/i
```
```suggestion
  const baseRegex = 'scroll|overlay|auto'
  const overflowScrollReg = includeHidden
    ? new RegExp(`${baseRegex}|hidden`, 'i')
     : new RegExp(baseRegex, 'i')
```

With this, we can include visually hidden elements, which I think it's a good addition. ``includeHidden`` must be an option.

ferferga replied:
@yue1123 You forgot about this one or you don't agree with me in the need for this?

ferferga replied:
@yue1123 Yes, exactly. I'm not 100% sure of the use case for this as well, but the virtual scroll library I'm using [takes this into account](https://github.com/rocwang/vue-virtual-scroll-grid/blob/main/src/utilites.ts#L98), [jQuery](https://github.com/fis-components/jquery-ui/blob/90fe5b0943d49c2369ad43fe9d6dba26f63f410f/core.js#L42) does as well

yue1123 replied:
@ferferga OK, I see. You’ve thought it through thoroughly — this situation can definitely happen, even if it’s a bit of an edge case. I’ll add an option to control it later, as you suggested.

yue1123 replied:
@ferferga If there’s any issue with the code in the PR, please point it out and I’ll make adjustments promptly, thank you.

ferferga commented on packages/core/useScrollParent/index.ts:
```diff
@@ -0,0 +1,81 @@
+    let node: ScrollParent = element
+    let level = 0
+    while (node && node !== rootScrollingElement && isElement(node as Element) && level < lookupLevel) {
+      const { overflowY } = window.getComputedStyle(node as Element)
+      if (overflowScrollReg.test(overflowY)) {
```
What if we're looking in the x axis? Why we can't return an object like { x, y } from the composable to be usable for those cases as well? So the user consumes whichever node it needs.

ferferga commented on packages/core/useScrollParent/index.ts:
```diff
@@ -0,0 +1,81 @@
+    return (
+      node.tagName !== 'HTML'
+      && node.tagName !== 'BODY'
+      && node.nodeType === ELEMENT_NODE_TYPE
```
```suggestion
      && node.nodeType === 1
```

I think this unnecessary assignment can be removed.

ferferga reviewed (commented):
Great, much cleaner, thank you! 🎉
Some other points you might have not come across!

Please, also update the tests to take into account the new regexes and x axis.

yue1123 committed:
feat(useScrollParent): improve code and add x-axis support

- Improved types
- Full coverage test cases
- Improved element validation and fallback handling

yue1123 commented:
Apologies for the delay! I was focused on refining the test cases. Really appreciate your detailed code review and thoughtful suggestions!

yue1123 commented:
Apologies，I forgot to submit my comments.🤣

yue1123 committed:
feat(useScrollParent): improve edge case handling

1. add includeHidden optional
2. support for fixed-position elements
3. Ensured absolute-positioned elements correctly skip static parent

ferferga reviewed (approved):
LGTM! 🚀

yue1123 committed:
Merge branch 'vueuse:main' into feat/useScrollParent

OrbisK committed:
Merge branch 'main' into feat/useScrollParent
